### PR TITLE
[FSTORE-332] implement method for get_or_create_feature_view

### PR DESCRIPTION
**Pull request by davitbzh**
This PR adds/fixes/changes...
- added get_or_create_feature_view method

JIRA Issue: -

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [x] Unit Tests
- [ ] Integration Tests
- [x] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: python/hsfs/feature_store.py [hunk 1]
```diff
@@ -963,6 +963,58 @@ def create_feature_view(
...
+                transformations.
+
+        # Returns:
+            `FeatureView`: The feature view metadata object.
+        """
+
+        try:
+            return self._feature_view_engine.get(name, version)
+        except exceptions.RestAPIError as e:
+            if (
```
Comment (kennethmhc) [inline python/hsfs/feature_store.py:1001]: can reuse `create_feature_view` here.

File: java/src/main/java/com/logicalclocks/hsfs/FeatureStore.java [hunk 2]
```diff
@@ -338,6 +342,38 @@ public FeatureView.FeatureViewBuilder createFeatureView() {
...
+   * @param name name of the feature view
+   * @param query Query object
+   * @param version version of the feature view
+   * @param description description of the feature view
+   * @param labels list of label features
+   * @return FeatureView
+   * @throws FeatureStoreException
+   * @throws IOException
+   */
+  public FeatureView getOrcreateFeatureView(String name, Query query, Integer version, String description,
```
Comment (kennethmhc) [inline java/src/main/java/com/logicalclocks/hsfs/FeatureStore.java:372]:
```suggestion
  public FeatureView getOrCreateFeatureView(String name, Query query, Integer version, String description,
```

File: java/src/main/java/com/logicalclocks/hsfs/FeatureStore.java [hunk 3]
```diff
@@ -338,6 +342,38 @@ public FeatureView.FeatureViewBuilder createFeatureView() {
...
+   * Get feature view metadata object or create a new one if it doesn't exist. This method doesn't update
+   * existing feature view metadata object.
+   *
+   * @param name name of the feature view
+   * @param version version of the feature view
+   * @return FeatureView
+   * @throws FeatureStoreException
+   * @throws IOException
+   */
+  public FeatureView getOrcreateFeatureView(String name, Integer version) throws FeatureStoreException, IOException {
```
[the remark this answers is not in the document]
Comment (kennethmhc) [reply java/src/main/java/com/logicalclocks/hsfs/FeatureStore.java:355]: `Query` is compulsory for creating a fv, need to include in the parameter as well.

File: java/src/main/java/com/logicalclocks/hsfs/metadata/FeatureViewApi.java [hunk 4]
```diff
@@ -119,6 +119,20 @@ public List<FeatureView> get(FeatureStore featureStore, String name) throws Feat
     }
   }
 
+  public FeatureView getOrCreateFeatureView(FeatureStore featureStore, String name, Integer version,  Query query,
```
Comment (kennethmhc) [inline java/src/main/java/com/logicalclocks/hsfs/metadata/FeatureViewApi.java:122]: This should be in `FeatureViewEngine`

**Review by kennethmhc — CHANGES_REQUESTED** [review]

File: java/src/main/java/com/logicalclocks/hsfs/engine/FeatureViewEngine.java [hunk 5]
```diff
@@ -398,4 +398,17 @@ public Map<String, Object> getTags(FeatureView featureView, Integer trainingData
...
   }
+
+  public FeatureView getOrCreateFeatureView(FeatureStore featureStore, String name, Integer version,  Query query,
+                                            String description, List<String> labels) {
+    FeatureView featureView = null;
+    try {
+      featureView = get(featureStore, name, version);
+    } catch (IOException | FeatureStoreException e) {
+      if (e.getMessage().contains("Error: 404") && e.getMessage().contains("\"errorCode\":270181")) {
+        featureView = new FeatureView(name, version, query, description, featureStore, labels);
```
Comment (kennethmhc) [inline java/src/main/java/com/logicalclocks/hsfs/engine/FeatureViewEngine.java:409]: This does not save fv in the backend, either using a builder or move this part to `FeatureStore` and reuse `FeatureStore.createFeatureView`.

**Review by kennethmhc — CHANGES_REQUESTED** [review]

File: python/hsfs/feature_store.py [hunk 6]
```diff
@@ -963,6 +963,56 @@ def create_feature_view(
         )
         return self._feature_view_engine.save(feat_view)
 
+    def get_or_create_feature_view(
+        self,
+        name: str,
+        query: Query,
+        version: int,
```
[the remark this answers is not in the document]
Comment (davitbzh) [reply python/hsfs/feature_store.py:970]: This is to be compatible with `get_or_create_feature_group`. There version is required. we agreed that  for `get_or_create` methods versions are required

File: java/src/main/java/com/logicalclocks/hsfs/engine/FeatureViewEngine.java [hunk 7]
```diff
@@ -400,13 +400,19 @@ public Map<String, Object> getTags(FeatureView featureView, Integer trainingData
...
+                                            String description, List<String> labels)
+      throws FeatureStoreException, IOException {
     FeatureView featureView = null;
     try {
       featureView = get(featureStore, name, version);
     } catch (IOException | FeatureStoreException e) {
       if (e.getMessage().contains("Error: 404") && e.getMessage().contains("\"errorCode\":270181")) {
-        featureView = new FeatureView(name, version, query, description, featureStore, labels);
+        featureView = new FeatureView.FeatureViewBuilder(featureStore)
+            .name(name)
```
Comment (kennethmhc) [inline java/src/main/java/com/logicalclocks/hsfs/engine/FeatureViewEngine.java:411]: `description` is missing.

Comment (davitbzh) [reply java/src/main/java/com/logicalclocks/hsfs/engine/FeatureViewEngine.java:411]: ahh, thanks. good catch

**Review by kennethmhc — APPROVED** [review]
LGTM!

**Pull request merged**